File: island_sim/src/prediction/postprocess.py
"""Post-processing: static cell overrides, smart probability floor, normalization."""

from __future__ import annotations

import numpy as np

from ..constants import (
    CLASS_EMPTY,
    CLASS_MOUNTAIN,
    CLASS_PORT,
    FOREST,
    MOUNTAIN,
    NUM_CLASSES,
    OCEAN,
)
from ..types import WorldState

# Probability floor to prevent catastrophic KL divergence
PROB_FLOOR = 0.01


def postprocess(
    prediction: np.ndarray,
    initial_state: WorldState,
) -> np.ndarray:
    """Apply smart probability floor and static cell overrides.

    Key insight: only apply probability floor to classes that are actually
    possible for each cell. Mountains never appear on non-mountain cells,
    ports only appear on coastal cells, etc.
    """
    result = prediction.copy()
    grid = initial_state.grid

    # Precompute coastal mask (adjacent to ocean)
    coastal_mask = np.zeros((initial_state.height, initial_state.width), dtype=bool)
    for y in range(initial_state.height):
        for x in range(initial_state.width):
            if grid[y, x] in (OCEAN, MOUNTAIN):
                continue
            for dy in range(-1, 2):
                for dx in range(-1, 2):
                    ny, nx = y + dy, x + dx
                    if 0 <= ny < initial_state.height and 0 <= nx < initial_state.width:
                        if grid[ny, nx] == OCEAN:
                            coastal_mask[y, x] = True

    for y in range(initial_state.height):
        for x in range(initial_state.width):
            terrain = int(grid[y, x])

            if terrain == OCEAN:
                result[y, x] = _make_exact_static(CLASS_EMPTY)
                continue
            if terrain == MOUNTAIN:
                result[y, x] = _make_exact_static(CLASS_MOUNTAIN)
                continue

            # Dynamic cell: apply floor only to plausible classes
            # Class 5 (mountain) is NEVER possible on non-mountain cells
            result[y, x, CLASS_MOUNTAIN] = 0.0

            # Class 2 (port) is only possible on coastal cells
            if not coastal_mask[y, x]:
                result[y, x, CLASS_PORT] = 0.0

            # Apply floor to remaining classes
            for cls in range(NUM_CLASSES):
                if result[y, x, cls] > 0.0 or cls in (CLASS_EMPTY, 1, 3, 4):  # empty, settlement, ruin, forest always possible
                    if cls == CLASS_MOUNTAIN:
                        continue  # already zeroed
                    if cls == CLASS_PORT and not coastal_mask[y, x]:
                        continue  # already zeroed
                    result[y, x, cls] = max(result[y, x, cls], PROB_FLOOR)

            # Renormalize
            row_sum = result[y, x].sum()
            if row_sum > 1e-10:
                result[y, x] /= row_sum

    return result


def _make_exact_static(dominant_class: int) -> np.ndarray:
    """Create a probability vector with 1.0 for the dominant class, 0.0 elsewhere."""
    dist = np.zeros(NUM_CLASSES)
    dist[dominant_class] = 1.0
    return dist

File: island_sim/rust/src/postprocess.rs
/// Post-processing: smart probability floor + static overrides.
/// Matches Python src/prediction/postprocess.py.

use crate::types::*;

// Two-tier probability floor: near settlements vs remote
const PROB_FLOOR: f64 = 0.005;
const PROB_FLOOR_REMOTE: f64 = 0.001;
const SETTLEMENT_PROXIMITY: i32 = 6;

const CLASS_EMPTY: usize = 0;
const CLASS_SETTLEMENT: usize = 1;
const CLASS_PORT: usize = 2;
const CLASS_RUIN: usize = 3;
const CLASS_FOREST: usize = 4;
const CLASS_MOUNTAIN: usize = 5;

/// Check if cell (x, y) is adjacent to ocean.
fn is_coastal(grid: &[Vec<i32>], x: usize, y: usize, height: usize, width: usize) -> bool {
    for dy in -1i32..=1 {
        for dx in -1i32..=1 {
            let ny = y as i32 + dy;
            let nx = x as i32 + dx;
            if ny >= 0 && ny < height as i32 && nx >= 0 && nx < width as i32 {
                if grid[ny as usize][nx as usize] == OCEAN {
                    return true;
                }
            }
        }
    }
    false
}

/// Apply postprocessing to a flat prediction array (H*W*6).
pub fn postprocess(prediction: &[f64], initial_state: &WorldState, height: usize, width: usize) -> Vec<f64> {
    let mut result = prediction.to_vec();

    // Precompute near-settlement mask
    let mut near_settlement = vec![vec![false; width]; height];
    for s in &initial_state.settlements {
        if !s.alive { continue; }
        for dy in -SETTLEMENT_PROXIMITY..=SETTLEMENT_PROXIMITY {
            for dx in -SETTLEMENT_PROXIMITY..=SETTLEMENT_PROXIMITY {
                let ny = s.y + dy;
                let nx = s.x + dx;
                if ny >= 0 && ny < height as i32 && nx >= 0 && nx < width as i32 {
                    near_settlement[ny as usize][nx as usize] = true;
                }
            }
        }
    }

    for y in 0..height {
        for x in 0..width {
            let terrain = initial_state.grid[y][x];
            let base = (y * width + x) * NUM_CLASSES;

            if terrain == OCEAN {
                make_exact_static(&mut result[base..base + NUM_CLASSES], CLASS_EMPTY);
                continue;
            }
            if terrain == MOUNTAIN {
                make_exact_static(&mut result[base..base + NUM_CLASSES], CLASS_MOUNTAIN);
                continue;
            }

            // Dynamic cell: zero out impossible classes
            result[base + CLASS_MOUNTAIN] = 0.0;

            let coastal = is_coastal(&initial_state.grid, x, y, height, width);
            if !coastal {
                result[base + CLASS_PORT] = 0.0;
            }

            let is_near = near_settlement[y][x];

            // Apply floor to plausible classes
            for cls in 0..NUM_CLASSES {
                if cls == CLASS_MOUNTAIN { continue; }
                if cls == CLASS_PORT && !coastal { continue; }

                // Settlement, Port, Ruin use remote floor when far from settlements
                let floor = if (cls == CLASS_SETTLEMENT || cls == CLASS_PORT || cls == CLASS_RUIN) && !is_near {
                    PROB_FLOOR_REMOTE
                } else {
                    PROB_FLOOR
                };

                if result[base + cls] < floor {
                    result[base + cls] = floor;
                }
            }

            // Normalize
            let sum: f64 = result[base..base + NUM_CLASSES].iter().sum();
            if sum > 1e-10 {
                for cls in 0..NUM_CLASSES {
                    result[base + cls] /= sum;
                }
            }
        }
    }

    result
}

fn make_exact_static(dist: &mut [f64], dominant_class: usize) {
    for c in 0..NUM_CLASSES {
        dist[c] = if c == dominant_class { 1.0 } else { 0.0 };
    }
}
